test/micro/org/openjdk/bench/vm/compiler/SparseCodeCache.java

```diff
@@ -0,0 +1,311 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+package org.openjdk.bench.vm.compiler;
+
+import java.lang.reflect.Method;
+import java.util.Random;
+
+import java.util.concurrent.TimeUnit;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Level;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+
+import org.openjdk.bench.util.InMemoryJavaCompiler;
+
+import jdk.test.whitebox.WhiteBox;
+import jdk.test.whitebox.code.NMethod;
+
+/*
+ * This benchmark is used to check performance when the code cache is sparse.
+ *
+ * We use C2 compiler to compile the same Java method multiple times
+ * to produce as many code as needed.
+ * These compiled methods represent the active methods in the code cache.
+ * We split active methods into groups.
+ * We put a group into a fixed size code region.
+ * We make a code region size aligned.
+ * CodeCache becomes sparse when code regions are not fully filled.
+ *
+ * The benchmark parameters are active method count, group count, and code region size.
+ */
+
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+@State(Scope.Benchmark)
+@Fork(value = 3, jvmArgs = {
+    "-XX:+UnlockDiagnosticVMOptions",
+    "-XX:+UnlockExperimentalVMOptions",
+    "-XX:+WhiteBoxAPI",
+    "-Xbootclasspath/a:lib-test/wb.jar",
+    "-XX:CompileCommand=dontinline,A::sum",
+    "-XX:-UseCodeCacheFlushing",
+    "-XX:-TieredCompilation",
+    "-XX:+SegmentedCodeCache",
+    "-XX:ReservedCodeCacheSize=320m",
+    "-XX:InitialCodeCacheSize=320m",
+    "-XX:+UseSerialGC",
+    "-XX:+PrintCodeCache"
+})
+public class SparseCodeCache {
+
+    static byte[] num1;
+    static byte[] num2;
+
+    @State(Scope.Thread)
+    public static class ThreadState {
+        byte[] result;
+
+        @Setup
+        public void setup() {
+            result = new byte[num1.length + 1];
+        }
+    }
+
+    private static Object WB;
+
+    @Param({"128", "256", "512", "768", "1024"})
```

Comment:
Big number of parameters is good for research purposes, but I think we should limit it for those people who run full microbenchmarks set to find regressions.